src/dbnode/storage/shard.go

```diff
@@ -1766,15 +1767,24 @@ func (s *dbShard) FetchBlocksMetadataV2(
 		}
 
 		// Return the reader to the cache
-		endPos := int64(reader.MetadataRead())
-		s.namespaceReaderMgr.put(reader)
+		err = s.namespaceReaderMgr.put(reader)
+		if err != nil {
+			return nil, nil, err
+		}
 
 		if numResults >= limit {
+			endPos := int64(reader.MetadataRead())
```

Comment:
I think `endPos` and `volume` need to be saved before putting the reader back in the cache. Checking them after is undefined.